File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java
/****************************************************************
 * Licensed to the Apache Software Foundation (ASF) under one   *
 * or more contributor license agreements.  See the NOTICE file *
 * distributed with this work for additional information        *
 * regarding copyright ownership.  The ASF licenses this file   *
 * to you under the Apache License, Version 2.0 (the            *
 * "License"); you may not use this file except in compliance   *
 * with the License.  You may obtain a copy of the License at   *
 *                                                              *
 *   http://www.apache.org/licenses/LICENSE-2.0                 *
 *                                                              *
 * Unless required by applicable law or agreed to in writing,   *
 * software distributed under the License is distributed on an  *
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
 * KIND, either express or implied.  See the License for the    *
 * specific language governing permissions and limitations      *
 * under the License.                                           *
 ****************************************************************/

package org.apache.james.jmap.model.mailbox;

import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.function.BinaryOperator;

import org.apache.james.mailbox.model.MailboxACL;
import org.apache.james.mailbox.model.MailboxACL.EntryKey;
import org.apache.james.mailbox.model.MailboxACL.NameType;
import org.apache.james.mailbox.model.MailboxACL.Rfc4314Rights;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.fasterxml.jackson.annotation.JsonAnyGetter;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import com.github.fge.lambdas.Throwing;
import com.github.steveash.guavate.Guavate;
import com.google.common.collect.ArrayListMultimap;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Multimap;

public class Rights {
    public enum Right {
        Administer('a', MailboxACL.Right.Administer),
        Expunge('e', MailboxACL.Right.PerformExpunge),
        Insert('i', MailboxACL.Right.Insert),
        Lookup('l', MailboxACL.Right.Lookup),
        Read('r', MailboxACL.Right.Read),
        Seen('s', MailboxACL.Right.WriteSeenFlag),
        T_Delete('t', MailboxACL.Right.DeleteMessages),
        Write('w', MailboxACL.Right.Write);

        private final char imapRight;
        private final MailboxACL.Right right;

        Right(char imapRight, MailboxACL.Right right) {
            this.imapRight = imapRight;
            this.right = right;
        }

        @JsonValue
        public char getImapRight() {
            return imapRight;
        }

        public MailboxACL.Right getRight() {
            return right;
        }

        public static boolean exists(char c) {
            return Arrays.stream(values())
                .anyMatch(right -> right.getImapRight() == c);
        }

   //     @JsonCreator
        public static Right forChar(char c) {
            return Arrays.stream(values())
                .filter(right -> right.getImapRight() == c)
                .findAny()
                .orElseThrow(() -> new IllegalArgumentException("No matching right for '" + c + "'"));
        }
    }

    public static class Username {
        private final String value;

     //   @JsonCreator
        public Username(String value) {
            this.value = value;
        }

        @JsonValue
        public String getValue() {
            return value;
        }

        @Override
        public final boolean equals(Object o) {
            if (o instanceof Username) {
                Username username = (Username) o;

                return Objects.equals(this.value, username.value);
            }
            return false;
        }

        @Override
        public final int hashCode() {
            return Objects.hash(value);
        }
    }

    public static class Builder {
        private Multimap<Username, Right> rights;

        public Builder() {
            rights = ArrayListMultimap.create();
        }

        public Builder delegateTo(Username username, Right... rights) {
            delegateTo(username, Arrays.asList(rights));
            return this;
        }

        public Builder delegateTo(Username username, Collection<Right> rights) {
            this.rights.putAll(username, rights);
            return this;
        }

        public Builder combine(Builder builder) {
            this.rights.putAll(builder.rights);
            return this;
        }

        public Rights build() {
            return new Rights(rights);
        }
    }

    public static Builder builder() {
        return new Builder();
    }

    public static Rights fromACL(MailboxACL acl) {
        return acl.getEntries()
            .entrySet()
            .stream()
            .filter(entry -> isUser(entry.getKey()))
            .map(entry -> builder().delegateTo(
                new Username(entry.getKey().getName()),
                fromACL(entry.getValue())))
            .reduce(builder(), Builder::combine)
            .build();
    }

    private static List<Right> fromACL(Rfc4314Rights rights) {
        return rights.list()
            .stream()
            .map(MailboxACL.Right::asCharacter)
            .filter(Rights::existingChar)
            .map(Right::forChar)
            .collect(Guavate.toImmutableList());
    }

    private static boolean existingChar(Character c) {
        if (!Right.exists(c)) {
            LOGGER.warn("Non handled right '" + c + "'");
            return false;
        }
        return true;
    }

    private static boolean isUser(EntryKey key) {
        if (key.isNegative()) {
            LOGGER.warn("Negative keys are not supported");
            return false;
        }
        if (key.getNameType() != NameType.user) {
            LOGGER.warn(key.getNameType() + " is not sopported. Onlu 'user' is.");
            return false;
        }
        return true;
    }

    public static final Rights EMPTY = new Rights(ArrayListMultimap.create());

    private static final Logger LOGGER = LoggerFactory.getLogger(Rights.class);

    private final Multimap<Username, Right> rights;

    @JsonCreator
    public Rights(Multimap<Username, Right> rights) {
        this.rights = rights;
    }

    @JsonAnyGetter
    public Map<Username, Collection<Right>> getRights() {
        return rights.asMap();
    }

    public MailboxACL toMailboxAcl() {
        BinaryOperator<MailboxACL> union = Throwing.binaryOperator(MailboxACL::union);

        return rights.asMap()
            .entrySet()
            .stream()
            .map(entrie -> new MailboxACL(
                ImmutableMap.of(
                    EntryKey.createUser(entrie.getKey().value),
                    toMailboxAclRights(entrie.getValue()))))
            .reduce(new MailboxACL(), union);
    }

    private Rfc4314Rights toMailboxAclRights(Collection<Right> rights) {
        BinaryOperator<Rfc4314Rights> union = Throwing.binaryOperator(Rfc4314Rights::union);

        return rights.stream()
            .map(Right::getRight)
            .map(Throwing.function(Rfc4314Rights::new))
            .reduce(new Rfc4314Rights(), union);
    }

    @Override
    public final boolean equals(Object o) {
        if (o instanceof Rights) {
            Rights that = (Rights) o;

            return Objects.equals(this.rights, that.rights);
        }
        return false;
    }

    @Override
    public final int hashCode() {
        return Objects.hash(rights);
    }
}

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Rights.scala
/****************************************************************
 * Licensed to the Apache Software Foundation (ASF) under one   *
 * or more contributor license agreements.  See the NOTICE file *
 * distributed with this work for additional information        *
 * regarding copyright ownership.  The ASF licenses this file   *
 * to you under the Apache License, Version 2.0 (the            *
 * "License"); you may not use this file except in compliance   *
 * with the License.  You may obtain a copy of the License at   *
 *                                                              *
 * http://www.apache.org/licenses/LICENSE-2.0                   *
 *                                                              *
 * Unless required by applicable law or agreed to in writing,   *
 * software distributed under the License is distributed on an  *
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
 * KIND, either express or implied.  See the License for the    *
 * specific language governing permissions and limitations      *
 * under the License.                                           *
 * ***************************************************************/

package org.apache.james.jmap.mail

import org.apache.james.core.Username
import org.apache.james.mailbox.model.MailboxACL
import org.apache.james.mailbox.model.MailboxACL.{EntryKey, Rfc4314Rights}
import org.slf4j.{Logger, LoggerFactory}

import scala.collection.mutable
import scala.jdk.CollectionConverters._

object Right {
  val UNSUPPORTED: Option[Boolean] = None

  val Administer = Right(MailboxACL.Right.Administer)
  val Expunge = Right(MailboxACL.Right.PerformExpunge)
  val Insert = Right(MailboxACL.Right.Insert)
  val Lookup = Right(MailboxACL.Right.Lookup)
  val Read = Right(MailboxACL.Right.Read)
  val Seen = Right(MailboxACL.Right.WriteSeenFlag)
  val DeleteMessages = Right(MailboxACL.Right.DeleteMessages)
  val Write = Right(MailboxACL.Right.Write)

  private val allRights = Seq(Administer, Expunge, Insert, Lookup, Read, Seen, DeleteMessages, Write)

  def forRight(right: MailboxACL.Right): Option[Right] = allRights.find(_.right.equals(right))

  def forChar(c: Char): Option[Right] = allRights.find(_.asCharacter == c)
}

sealed case class Right(right: MailboxACL.Right) {
  val asCharacter: Char = right.asCharacter

  val toMailboxRight: MailboxACL.Right = right
}

object Rights {
  private val LOGGER: Logger = LoggerFactory.getLogger(classOf[Rights])

  val EMPTY = new Rights(Map())

  def of(username: Username, right: Right): Rights = of(username, Seq(right))

  def of(username: Username, rights: Seq[Right]): Rights = {
    require(rights.nonEmpty, "'rights' should not be empty")

    Rights(Map(username -> rights))
  }

  def fromACL(acl: MailboxACL): Rights = acl.getEntries.asScala
    .filter {
      case (entryKey, _) => isSupported(entryKey)
    }
    .map {
      case (entryKey, aclRights) => toRights(entryKey, aclRights)
    }
    .fold(EMPTY)(_ combine _)

  private def toRights(entryKey: MailboxACL.EntryKey, aclRights: MailboxACL.Rfc4314Rights): Rights =
    of(Username.of(entryKey.getName), fromACL(aclRights))

  private def fromACL(rights: MailboxACL.Rfc4314Rights): Seq[Right] = rights.list.asScala
      .toSeq
      .flatMap(Right.forRight)

  private def isSupported(key: MailboxACL.EntryKey): Boolean = {
    if (key.isNegative) {
      LOGGER.info("Negative keys are not supported")
      return false
    }
    if (key == MailboxACL.OWNER_KEY) {
      return false
    }
    if (key.getNameType ne MailboxACL.NameType.user) {
      LOGGER.info("{} is not supported. Only 'user' is.", key.getNameType)
      return false
    }
    true
  }
}

case class Rights private(rights: Map[Username, Seq[Right]]) {

  def removeEntriesFor(username: Username) = Rights(rights.filter(!_._1.equals(username)))

  def toMailboxAcl: MailboxACL = {
    val map: Map[EntryKey, Rfc4314Rights] = rights.view
      .mapValues(seq => toJavaRights(seq))
      .toMap
      .map {
        case (user, rfc4314Rights) => (EntryKey.createUserEntryKey(user), rfc4314Rights)
      }
    new MailboxACL(map.asJava)
  }

  def append(username: Username, right: Right): Rights = append(username, Seq(right))

  def append(username: Username, rights: Seq[Right]): Rights = {
    require(rights.nonEmpty, "'rights' should not be empty")

    Rights(this.rights + (username -> rights))
  }

  def combine(that: Rights): Rights = Rights(this.rights ++ that.rights)

  private def toJavaRights(seq: Seq[Right]): Rfc4314Rights = Rfc4314Rights.of(seq.map(_.right).asJava)

  def mayReadItems(username: Username): Option[Boolean] = containsRight(username, Right.Read)

  def mayAddItems(username: Username): Option[Boolean] = containsRight(username, Right.Insert)

  def mayRemoveItems(username: Username): Option[Boolean] = containsRight(username, Right.DeleteMessages)

  def mayCreateChild(username: Username): Option[Boolean] = Right.UNSUPPORTED

  def mayRename(username: Username): Option[Boolean] = Right.UNSUPPORTED

  def mayDelete(username: Username): Option[Boolean] = Right.UNSUPPORTED

  private def containsRight(username: Username, right: Right): Option[Boolean] = rights.get(username)
    .map(_.contains(right))
}
